File: src/services/discord/channels.ts
import { Client, Guild, Channel, GuildMember, Collection } from 'discord.js';

import { channelIds, guildId, moods } from '@/config/discord';
import { discordApi } from '@/services/discord/api';
import { getRandom, logger } from '@/utils';

export const updateBoosterCount: (bot: Client) => void = (bot: Client) => {
  try {
    const guild: Guild | undefined = bot.guilds.cache.get(guildId);
    if (!guild) return logger('ERROR', 'GUILD NOT FOUND', guildId);

    const boosters: Collection<string, GuildMember> = guild.members.cache.filter((member: GuildMember) => member.premiumSince);
    const boosterCount: number = boosters.size;

    const channel: Channel | undefined = guild.channels.cache.get(channelIds['booster-count']);
    if (!channel) return logger('ERROR', 'CHANNEL NOT FOUND - BOOSTER COUNT', channelIds['booster-count']);

    channel.setName(`💎・meow club: ${boosterCount}`);
  } catch (error) {
    logger('ERROR', 'DISCORD CHANNELS - UPDATE BOOSTER COUNT', error);
  }
};

export const updateMembersCount: (bot: Client) => void = (bot: Client) => {
  try {
    const guild: Guild | undefined = bot.guilds.cache.get(guildId);
    if (!guild) return logger('ERROR', 'GUILD NOT FOUND', guildId);

    const memberCount: number = guild.memberCount;

    const channel: Channel | undefined = guild.channels.cache.get(channelIds['members-count']);
    if (!channel) return logger('ERROR', 'CHANNEL NOT FOUND - MEMBERS COUNT', channelIds['members-count']);

    channel.setName(`😺・meows: ${memberCount}`);
  } catch (error) {
    logger('ERROR', 'DISCORD CHANNELS - UPDATE MEMBERS COUNT', error);
  }
};

export const updateServerMood: () => Promise<void> = async () => {
  try {
    const mood: string = getRandom(moods);

    await discordApi.updateChannelName(channelIds['server-mood'], mood);
  } catch (error) {
    logger('ERROR', 'DISCORD CHANNELS - UPDATE SERVER MOOD', error);
  }
};

File: internal/services/discord/channels.go
package discord

import (
	"math/rand"
	"strconv"

	"github.com/MatheusIshiyama/meowhaha/pkg/logger"
	"github.com/bwmarrin/discordgo"
)

const (
	guildID              string = "1386212819669880913"
	serverMoodChannelID  string = "1386219879673237571"
	memberCountChannelID string = "1386226043320537190"
	boosterChannelID     string = "1388422157301977209"
)

var (
	moods = []string{
		"😺・Stay Paw-sitive",
		"🐾・Keep Purring",
		"😼・Claws Out",
		"🧶・Chase the Yarn",
		"😻・Dream Big, Meow",
		"🐱・Leap Higher",
		"🙀・Fearless Feline",
		"🐟・Catch Your Goal",
		"😽・Believe & Purr",
		"🌟・Shine, Little Cat",
	}
)

type ChannelService struct{}

func GetGuild(session *discordgo.Session) *discordgo.Guild {
	var Guild *discordgo.Guild

	for _, guild := range session.State.Guilds {
		if guild.ID == guildID {
			Guild = guild
			break
		}
	}

	if Guild == nil {
		logger.Error("DISCORD", "Guild not found", nil)
		return nil
	}

	return Guild
}

func (ChannelService) GetRandomMood() string {
	return moods[rand.Intn(len(moods))]
}

func (ChannelService) UpdateServerMood(session *discordgo.Session) {
	Guild := GetGuild(session)

	mood := Channel.GetRandomMood()

	channels := Guild.Channels
	for _, channel := range channels {
		if channel.ID == serverMoodChannelID {
			session.ChannelEditComplex(channel.ID, &discordgo.ChannelEdit{
				Name: mood,
			})
		}
	}
}

func (ChannelService) UpdateCounts(session *discordgo.Session) {
	Guild := GetGuild(session)

	boosterCount := 0
	for _, member := range Guild.Members {
		if member.PremiumSince != nil {
			boosterCount++
		}
	}

	channels := Guild.Channels
	for _, channel := range channels {
		if channel.ID == memberCountChannelID {
			session.ChannelEditComplex(channel.ID, &discordgo.ChannelEdit{
				Name: "😺・meows: " + strconv.Itoa(len(Guild.Members)),
			})
		}
		if channel.ID == boosterChannelID {
			session.ChannelEditComplex(channel.ID, &discordgo.ChannelEdit{
				Name: "💎・meow club: " + strconv.Itoa(boosterCount),
			})
		}
	}
}

var Channel = ChannelService{}
